gzip_io.ml

```diff
@@ -36,6 +36,31 @@ let string s =
   IO.nwrite out (Bytes.unsafe_of_string s); (* IO wrong type *)
   IO.close_out out
 
+let string_lwt ?(chunk_size = 3000) ?(yield = Lwt.pause) s =
+  let out = output (IO.output_string ()) in
+  let buff = Buffer.create chunk_size in
+  let len = String.length s in
+  let rec loop i =
+    if i >= len then (
+      (* Final flush of the buffer if there's any residue *)
+      if Buffer.length buff > 0 then IO.nwrite out (Buffer.to_bytes buff);
+      Lwt.return_unit)
+    else begin
+      let c = s.[i] in
+      Buffer.add_char buff c;
+      if Buffer.length buff < chunk_size then loop (i + 1)
+      else (
+        (* Buffer is full, write and clear it *)
+        IO.nwrite out (Buffer.to_bytes buff);
+        Buffer.clear buff;
+        (* Yield after processing a chunk *)
+        let%lwt () = yield () in
+        loop (i + 1))
+    end
+  in
+  let%lwt () = loop 0 in
```

Comment:
There's no need to process the input string char by char (slow!), accumulate the data to be compressed in a buffer, then allocate a string corresponding to the data (ouch!). It should be possible to write substrings (not allocated, but with offset + size) from the original string to the IO.

Reply:
Just checked
https://github.com/ygrek/ocaml-extlib/blob/0104e03b77254454a4c50859286cadd1803d2490/src/IO.mli#L96-L105

Reply:
Thanks, I updated the code in [`ba084d4` (#25)](https://github.com/ahrefs/devkit/pull/25/commits/ba084d430e3c93d1bf918b347ac7efadad8acf4b)